Override environment for jobstart

teto opened the pull request:
This is very raw attempt at https://github.com/neovim/neovim/issues/4970.

teto commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11687,61 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+            // int res = sprintf(env[count-1], "%s=%s", var->di_key, str));
+            // TODO need to escape the var 
+            // use shell_xescape_xquote ?
+            env[count-1] = (char *)concat_str(var->di_key, concat_str((char_u *)"=", (char_u*)(str)));
```
2 questions here:
- should the conversion of {key:value} into ["key=value"] be moved to tv_to_argv ?
- the test succeeds without the space in `{'TOTO': 'hello world'}`. I believe I should use shell_xescape_xquote (via removing its static qualifier) ?

jamessan replied:
Why is any escaping required when constructing the environment?  If anything, the consumer of the environment variable should be ready to deal with whitespace and the like, but as far as I know when setting you just need `key=var` format.

teto replied:
hum I think you are right. I thought of that because I have failures when running `TEST_FILE=test/functional/core/job_spec.lua TEST_TAG=env make functionaltest`
I get
```
Passed in:
(table) {
  [1] = 'notification'
 *[2] = 'stderr'
  [3] = {
    [1] = 0
    [2] = {
      [1] = '/bin/zsh: can't open input file: '
      [2] = '' } } }
Expected:
(table) {
  [1] = 'notification'
 *[2] = 'stdout'
  [3] = {
    [1] = 0
    [2] = {
      [1] = 'hello world'
      [2] = '' } } }

```
I am not sure why, I tried passing a list instead of a command (e.g., `jobstart(['echo', '$TOTO'], ...` to no avail.

jamessan replied:
`uv_spawn()` uses the specified environment as the _entire_ environment.  Since you haven't implemented the ability to augment the existing environment, then `jobstart()` is running without a `$PATH`.

bfredl commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11687,61 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+          if (str) {
+            ILOG("key %s", str);
+            if (env == NULL) {
+              env = xmalloc((size_t)1);
```
Use `sizeof(*env)` not 1

bfredl replied:
Also complete size can be precalculated

jamessan commented:
For reference, @bfredl and I had a discussion (starting [here](https://gitter.im/neovim/neovim?at=599598e176a757f808b43c3d) about the handling of the environment on Gitter.

The TL;DR version is basically:

* `jobstart(..., {'env': { ... }, ...})` should overlay the existing environment.  This means we need to know the full environment that _would_ be inherited if we weren't modifying the environment, since `uv_spawn` treats the specified environment as the exact environment that should be used for the child process.
* `jobstart(..., {'env': { ... }, 'reset_env': v:true, ...})` should use _only_ the environment specified.
  * This also provides the ability for future expansion on the semantics of `'reset_env'`, perhaps providing a value that means "build from a sane base state" (similar to what happens for services in *nix) instead of a clean environment.

The suggestions in #6765 for APIs and tracking the environment should be kept in mind.  I think we'll need at least to manage the environment to properly handle giving manipulated environments to child processes.  The APIs should be able to naturally fall out from that, but I wouldn't be opposed to having a follow up PR for those.

justinmk commented:
`inherit_env=false` feels more natural than `reset_env=true`.

bfredl commented:
Don't we have most Boolean flags defaulting to false?

teto commented:
As tv_dict_get_number(job_opts, "var") returns 0 if "var" is not set, it's best to have a "false" defaults, hence I added the "reset_env" version.

I still have a problem when overriding environment (hence no PATH so I use fullpath to binary) and trying to call `jobstart(['/usr/bin/env'])`. I wonder if libuv doesn't like that or if it's sthg else.
```
(table) {
  [1] = 'notification'
 *[2] = 'stderr'
  [3] = {
    [1] = 0
    [2] = {
      [1] = '/usr/bin/env: '': No such file or directory'
      [2] = '' } } }
Expected:
(table) {
  [1] = 'notification'
 *[2] = 'stdout'
  [3] = {
    [1] = 0
    [2] = {
      [1] = 'TOTO=hello world'
      [2] = '' } } }

```

Also it does not support values longer than NUMBUFLEN because of tv_get_string

justinmk commented:
> Don't we have most Boolean flags defaulting to false?

Not to my knowledge. Usually "positive logic" is preferred, I've never heard of a preference for the initial value. "reset" is negative logic, it's like naming a variable `disable_thing` instead of `enable_thing` or `not_condition` instead of `condition`. 

Also "inheriting the environment" is more familiar concept, isn't it?

> As tv_dict_get_number(job_opts, "var") returns 0 if "var" is not set, it's best to have a "false" defaults, hence I added the "reset_env" version.

`!!tv_dict_get_number(job_opts, "var")` ? Internal details should not decide the user interface...

justinmk commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      ILOG("%d passed env variables", env_size);
+
+      if (!reset_env) {
+        char **environ = os_getfullenv();
```
Now I get it. Here we actually take action to attempt to _gather_ the full environment and pass it on. That's rather different than what is implied: "reset_env" sounds like it takes action to _delete_ an environment that would otherwise be inherited.

Something like `include_parent_env` would be better, but too long ... so I'll be quiet.

bfredl commented:
I don't think flags should be polarized on "which one libuv requires us to do more work", rather on what for the _user_ is most action. "Adding a few env bars" is a less intrusive action than "completely resetting the env"

ZyX-I commented:
I would say that false default is fine, but I do not like `reset_env`. Suggestions: `ignore_env` (after `env` shell command), `clear_env` (there is a libc function `clearenv`).

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+        }
+        ILOG("current env_len=%d", current_env_size);
+        env_size += current_env_size;
+        env = xmalloc( (env_size) * sizeof(char));
```
`sizeof(char)` is guaranteed to be 1 by C99 standard itself. Also don’t use `sizeof(type)` as a general rule, use variables: in your case it would be `sizeof(*env)`, which is also size of the *pointer* `char *` like it should be, with `sizeof(char)` it should SEGV.

ZyX-I replied:
Also here are useless parenthesis around `env_size` and space after `(` which should not be present.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+        }
+      } else {
+        // + 1 for last null entry
+        env = xmalloc( (env_size + 1) * sizeof(char));
```
Same.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      size_t i = 0;
+      // size_t orig_env_size = 0;
+      if (item->di_tv.v_type != VAR_DICT) {
+        EMSG2(_(e_invarg2), "Expected a dictionary for 'env'");
```
It is better to use a new error number, never use `e_invarg` and avoid `e_invarg2`. And English phrases are supposed to be translatable.

justinmk replied:
@jamessan zyx's comment is still relevant. I would just change it to `EMSG2(e_invarg2, "env")` for now.

> way to choose a free error code ? or should I just search & increase an number until I can't find a match ?

https://github.com/neovim/neovim/pull/8155

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+          if (str) {
+            size_t len = STRLEN(var->di_key) + strlen(str) + strlen("=") + 1;
+            ILOG("value #%d: %s", i, str);
+            env[i] = xmallocz(len * sizeof(char));
```
Remove useless `sizeof(char)`. Or replace with `sizeof(env[0][0])`.

ZyX-I replied:
Also you allocated one more bytes then needed: +1 in `len` calculation is for NUL I guess, but `xmallocz` already allocates one more bytes for NUL. Also given that you use `snprintf` `xmallocz` is not needed, `snprintf` will add bytes itself, so just replace code with `env[i] = xmalloc(len);`.

teto commented on src/nvim/eval.c:
```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      size_t i = 0;
+      // size_t orig_env_size = 0;
+      if (item->di_tv.v_type != VAR_DICT) {
+        EMSG2(_(e_invarg2), "Expected a dictionary for 'env'");
```
The original intended usage may be more like `EMSG2(_(e_invarg2), "env");`. I hate the "invalid argument" message because it's so vague. So here I should use a new message like e_expectingdict. Is there an immediate way to choose a free error code ? or should I just search & increase an number until I can't find a match ?

jamessan commented:
> In my opinion it is better go with ignore_env, it closer represents what code actually does.

`clear_env`/`ignore_env` are similar to what I had originally suggested.  @bfredl suggested `reset_env` since it provides more room to expand functionality.  If that's not of interest, then I would go with `ignore_env` as well.

bfredl commented:
I never said extensibility is tied to the name "reset_env", it was just the first name that was  suggested. The point  was just the form "single dict + Boolean flag being false by default"

teto commented:
vim just uses "env" https://github.com/vim/vim/commit/05aafed54b50b602315ae55d83a7d089804cecb0

jamessan commented:
> vim just uses "env"

However, it doesn't specify whether that's additive or the entire environment.  If it's additive, I don't see a way to remove certain environment variables.

teto commented:
I pushed an update that for some reason crashes, always at the same position
```
INFO  2018-11-08T12:57:45.878 32468 f_jobstart:11803: resulting env (run 26) 0x7f7512bb2c00
INFO  2018-11-08T12:57:45.878 32468 f_jobstart:11804: resulting env (run 26) src=/nix/store/zg7vi1k141fkhimmbqs2l0z5zymx1gsq-source
INFO  2018-11-08T12:57:45.878 32468 f_jobstart:11803: resulting env (run 27) 0x7f7512a1f9f0
INFO  2018-11-08T12:57:45.878 32468 f_jobstart:11804: resulting env (run 27) ZSH_AUTOSUGGEST_STRATEGY=default
INFO  2018-11-08T12:57:45.878 32468 f_jobstart:11803: resulting env (run 28) 0x47554c505f5a4f4d

```
The address in position 28 get changed to a wrong value (0x47554c505f5a4f4d instead of sthg like 0x7f7512a1f9f0) I am not sure what I do wrong.
I tried running valgrind 
`valgrind --trace-children=yes make functionaltest TEST_FILE=test/functional/core/job_spec.lua TEST_TAG=env`
which finds errors
```
==15652== Command: /nix/store/dx2zr2ddylvi7rqasmcmmf05am40di9j-lua-5.1.5/bin/lua -e package.path="/build/.luarocks//share/lua/5.1/?.lua;/build/.luarocks//share/lua/5.1/?/init.lua;/nix/store/xlwv4p9gdyrfc3gvi90csjlf6a97zgcs-luarocks-2.4.4/share/lua/5.1/?.lua;/nix/store/xlwv4p9gdyrfc3gvi90csjlf6a97zgcs-luarocks-2.4.4/share/lua/5.1/?/init.lua;"..package.path;\ package.cpath="/build/.luarocks//lib/lua/5.1/?.so;/nix/store/xlwv4p9gdyrfc3gvi90csjlf6a97zgcs-luarocks-2.4.4/lib/lua/5.1/?.so;"..package.cpath -e local\ k,l,_=pcall(require,"luarocks.loader")\ _=k\ and\ l.add_context("busted","2.0.rc12-1") /nix/store/mvppx40idcj6rwiczdwki8cbx94l6s24-lua5.1-busted-2.0.rc12-1/lib/luarocks/rocks/busted/2.0.rc12-1/bin/busted --tags=env -v -o nvim --lua=/nix/store/pi0bpz92zlyjnl4p4kpgii8f97y23pqa-lua-5.1.5-env/bin/lua --lazy --helper=/home/teto/neovim/test/functional/preload.lua --lpath=/home/teto/neovim/build/?.lua --lpath=?.lua test/functional/core/job_spec.lua
==15652== 
==15653== Memcheck, a memory error detector
==15653== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==15653== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==15653== Command: /bin/sh -c 
==15653== 
==15653== Invalid free() / delete / delete[] / realloc()
==15653==    at 0x4C2E2D8: free (in /nix/store/xq4x05f8d2a6w2hpx42pgvc3bz18791c-valgrind-3.13.0/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15653==    by 0x46189F: unwind_frame_run_internal.constprop.3 (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15653==    by 0x461C13: run_unwind_frame (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15653==    by 0x481FF3: parse_and_execute (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15653==    by 0x41BB48: run_one_command (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15653==    by 0x41FCC7: main (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15653==  Address 0x42325e8 is in the brk data segment 0x4227000-0x4254fff
==15653== 
==15653== 
==15653== HEAP SUMMARY:
==15653==     in use at exit: 0 bytes in 0 blocks
==15653==   total heap usage: 2,395 allocs, 2,396 frees, 150,905 bytes allocated
==15653== 
==15653== All heap blocks were freed -- no leaks are possible
==15653== 
==15653== For counts of detected and suppressed errors, rerun with: -v
==15653== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
==15654== Memcheck, a memory error detector
==15654== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==15654== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==15654== Command: /bin/sh -c /nix/store/pi0bpz92zlyjnl4p4kpgii8f97y23pqa-lua-5.1.5-env/bin/lua\ /nix/store/mvppx40idcj6rwiczdwki8cbx94l6s24-lua5.1-busted-2.0.rc12-1/lib/luarocks/rocks/busted/2.0.rc12-1/bin/busted\ --ignore-lua\ --tags=env\ -v\ -o\ nvim\ --lua=/nix/store/pi0bpz92zlyjnl4p4kpgii8f97y23pqa-lua-5.1.5-env/bin/lua\ --lazy\ --helper=/home/teto/neovim/test/functional/preload.lua\ --lpath=/home/teto/neovim/build/?.lua\ --lpath=?.lua\ test/functional/core/job_spec.lua
==15654== 
==15654== Invalid read of size 1
==15654==    at 0x4A27B7: internal_free.isra.1 (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x495B0E: glob_pattern_p (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x49673D: glob_filename (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x462966: shell_glob_filename (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x45DAD1: expand_word_list_internal (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x41CB57: execute_simple_command (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x436578: execute_command_internal (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x482417: parse_and_execute (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x41BB48: run_one_command (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==    by 0x41FCC7: main (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/bin/bash)
==15654==  Address 0x5b054a8 is 8 bytes before a block of size 256 alloc'd
==15654==    at 0x4C2D06C: malloc (in /nix/store/xq4x05f8d2a6w2hpx42pgvc3bz18791c-valgrind-3.13.0/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15654==    by 0x4C2F21F: realloc (in /nix/store/xq4x05f8d2a6w2hpx42pgvc3bz18791c-valgrind-3.13.0/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15654==    by 0x49A9DA: xdupmbstowcs (in /nix/store/5igbdc1czdss7341r360648n14pkpp5r-bash-interactive-4.4-p23/
```
I am not sure if that's due to an environmental error or my fault.

justinmk commented:
@teto valgrind is finding bugs in `bash` with that invocation :) 

To use valgrind with a test, set VALGRIND=1 in the environment, like this:

    VALGRIND=1 TEST_FILE=test/functional/core/job_spec.lua TEST_TAG=env make functionaltest

See also 
https://github.com/neovim/neovim/blob/c4c74c3883aa3122c0c877ca8dd7b26beb5cc4aa/test/functional/helpers.lua#L53 . It is also mentioned in `test/README.md` .

jamessan commented on src/nvim/eval.c:
```diff
@@ -11780,7 +11781,7 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
         const char *str = tv_get_string(&var->di_tv);
         if (str) {
-          size_t len = STRLEN(var->di_key) + strlen(str) + strlen("=");
+          size_t len = STRLEN(var->di_key) + strlen(str) + strlen("=") + 1;
```
`xmallocz()` adjusts the allocation length to account for the NUL.  It looks like you're including that in the calculation here so that `snprintf()` has an accurate length to work with, in which case the `xmallocz()` should be changed to `xmalloc()`.

teto commented:
thanks, I forgot about the test/README.md (searched the wiki with 0 results on valgrind). I fixed the bug (still need to free memory etc) but first I want the new test to succed. 

I've seen the new `expect_msg_seq` and tried to use that but it fails at printing an error message and  
the test doesn't print all the output it seems ? When I try to run 
```
(ins)[nix-shell:~/neovim/build]$ TEST_FILE=test/functional/core/job_spec.lua TEST_TAG=env make functionaltest
[  0%] Built target printargs-test
[100%] Built target nvim
[100%] Built target shell-test
[100%] Built target tty-test
[----------] Global test environment setup.
[----------] Running tests from test/functional/core/job_spec.lua
[ RUN      ] jobs append environment #env: 9.19 ms OK
[ RUN      ] jobs replace environment #env: Pretty printing err2
{
  message = [[
Expected objects to be the same.
Passed in:
(table) {
 *[1] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = { ... more } } }
  [2] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = { ... more } } }
  [3] = {
    [1] = 'notification'
    [2] = 'exit'
    [3] = {
      [1] = 0
      [2] = 0 } } }
Expected:
(table) {
 *[1] = 'notification'
  [2] = 'stdout'
  [3] = {
    [1] = 0
    [2] = {
      [1] = 'TOTO=hello world'
      [2] = '' } } }]]
}
true
END of pretty printing
ERR
test/functional/helpers.lua:147: bad argument #4 to 'format' (string expected, got table)

stack traceback:
	test/functional/helpers.lua:147: in function 'cat_err'
	test/functional/helpers.lua:165: in function 'expect_msg_seq'
	test/functional/core/job_spec.lua:71: in function <test/functional/core/job_spec.lua:65>

[----------] 2 tests from test/functional/core/job_spec.lua (61.99 ms total)

[----------] Global test environment teardown.
------------------------------------------------------------------------------
$NVIM_LOG_FILE: /tmp/log
(last 10 lines)
INFO  2018-11-09T00:43:32.814 6725  mch_exit:152: Nvim exit: 0
INFO  2018-11-09T00:43:32.823 6732  main:590: starting main loop
INFO  2018-11-09T00:43:32.827 6732  f_jobstart:11750: 1 passed env variables
INFO  2018-11-09T00:43:32.827 6732  f_jobstart:11757: Clearing env
INFO  2018-11-09T00:43:32.828 6732  f_jobstart:11783: value #0: hello world
INFO  2018-11-09T00:43:32.828 6732  f_jobstart:11783: generated string #0: TOTO=hello world (len 17)
INFO  2018-11-09T00:43:32.828 6732  libuv_process_spawn:52: Calling uv_spawn with env=0x7f5ec03da1c0
INFO  2018-11-09T00:43:32.828 6732  free_env:284: freeing TOTO=hello world
INFO  2018-11-09T00:43:32.828 6732  channel_create_event:208: new channel 3 (?) : {"id": 3, "mode": "bytes", "stream": "job"}
INFO  2018-11-09T00:43:32.830 6732  on_process_exit:385: exited: pid=6733 status=0 stoptime=0
------------------------------------------------------------------------------
[==========] 2 tests from 1 test file ran. (62.22 ms total)
[  PASSED  ] 1 test.
[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional/core/job_spec.lua @ 65: jobs replace environment #env
test/functional/helpers.lua:147: bad argument #4 to 'format' (string expected, got table)

stack traceback:
	test/functional/helpers.lua:147: in function 'cat_err'
	test/functional/helpers.lua:165: in function 'expect_msg_seq'
	test/functional/core/job_spec.lua:71: in function <test/functional/core/job_spec.lua:65>


 1 ERROR
-- Output to stderr:
```
It seems to work when I run a similar code from nvim 
`call jobstart(["env"], {'on_stdout':{j,d,e->append(line('.'),d)}, 'env': { 'TOTO': "test"}, "clear_env": 1 })`

teto committed:
test: always pass a string to expect_msg_seq

Seems like pcall doesn't always return a string as a 2nd element of the tuple.

teto committed:
test: new test for setting environment

teto committed:
jobstart now supports env/clear_env

to modify the environment of the launched job.

teto commented:
Tests work locally. I renamed the 'reset_env' to "clear_env" as suggested by @ZyX-I as I felt it was indeed a bit clearer and closer to libc convention.
Might need more tests (empty strings) etc. 

Regarding the test infrastructure:
- seems like pcall doesn't return `bool, string ` but `bool, <results>` and I didn't dive too much but "eq" doesn't return a simple string, which explains `test/functional/helpers.lua:147: bad argument #4 to 'format' (string expected, got table)`. I've added a quick hack to fix it
- to fully dump the stdout notifications, set assert:set_parameter("TableFormatLevel", 10000) (Can't we set this globally/for every test ?).

justinmk commented on src/nvim/event/libuv_process.c:
```diff
@@ -41,7 +41,7 @@ int libuv_process_spawn(LibuvProcess *uvproc)
   uvproc->uvopts.exit_cb = exit_cb;
   uvproc->uvopts.cwd = proc->cwd;
-  uvproc->uvopts.env = NULL;  // Inherits the parent (nvim) env.
+  uvproc->uvopts.env = proc->env;
```
should keep the comment explaining what NULL does, or maybe add it to the docstring for `channel_job_start()`.

jamessan replied:
> should keep the comment explaining what NULL does, or maybe add it to the docstring for `channel_job_start()`.

I added a docstring for `channel_job_start()`, including this tidbit.

jamessan committed:
Add os_getfullenv_size/os_copyfullenv

justinmk commented:
@jamessan If it saves work, don't bother adding this feature to `termopen()`, it is redundant with `jobstart()` and will be deprecated (if I ever finish https://github.com/neovim/neovim/pull/5485).

bfredl commented:
@justinmk I don't think most of #5485 is needed for that actually. `nvim_get_chan_info(job).buffer` already works to get the terminal buffer id. The only thing that is missing is a flag to `jobstart()` to allocate a _new_ buffer as the terminal buffer, instead of implicitly take over current window/buffer.  Then the caller can use `nvim_win_set_buf`, open a split, or a float, or whatever they want.

justinmk commented on src/nvim/channel.c:
```diff
@@ -277,11 +277,44 @@ static void close_cb(Stream *stream, void *data)
+/// @param[in]  pty_height  Height of the pty, ignored if `pty` is false
+/// @param[in]  term_name  `$TERM` for the pty
+/// @param[in]  env  Nvim's configured environment is used if this is NULL,
+///                  otherwise defines all environment variables
```
freed on return

justinmk commented on src/nvim/channel.c:
```diff
@@ -277,11 +277,44 @@ static void close_cb(Stream *stream, void *data)
+/// @param[in]  detach  True if the job should not be killed when nvim exits,
+///                     ignored if `pty` is true
+/// @param[in]  cwd  Initial working directory for the job.  Nvim's working
+///                  directory if `cwd` is NULL
```
freed on return

justinmk commented on src/nvim/channel.c:
```diff
@@ -333,6 +367,7 @@ Channel *channel_job_start(char **argv, CallbackReader on_stdout,
   if (status) {
     EMSG3(_(e_jobspawn), os_strerror(status), cmd);
     xfree(cmd);
+    free_env(proc->env);
```
`free_env(env)` is consistent with the `cwd` treatment.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -32,6 +32,9 @@ describe('jobs', function()
   before_each(function()
     clear()
+
+    assert:set_parameter('TableFormatLevel', 1000000)
```
Ah didn't know about this. That's why I added this: https://github.com/neovim/neovim/commit/3d1ed7c959017dc8664497b26c86c9ffabf02891#diff-75e5cc780c1d61f0c237fc0c0055711bR431

janlazo commented:
Is this blocked?

jamessan commented:
> Is this blocked?

Windows side of things needs to be fixed.

jamessan commented:
Ok, this should be ready now.

justinmk commented:
1 failing test:

```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] test/functional\ui\wildmode_spec.lua @ 295: command line completion completes env var names #9681
test/functional\ui\wildmode_spec.lua:301: Row 1 did not match.
Expected:
  |*                                        |
  |{1:~                                       }|
  |{1:~                                       }|
  |*{2:XTEST_1}{3:  XTEST_2                        }|
  |*:!echo $XTEST_1^                         |
Actual:
  |*^                                        |
  |{1:~                                       }|
  |{1:~                                       }|
  |*{1:~                                       }|
  |*                                        |
```

jamessan committed:
os_getenvname_at_index: Handle Windows env vars whose name starts with =

jamessan commented:
> 1 failing test:

Fixed

teto commented on src/nvim/channel.c:
```diff
@@ -272,11 +272,44 @@ static void close_cb(Stream *stream, void *data)
+{
+  if (!env) { return; }
+  for (char **it = env; *it; it++) {
+    xfree(*it);
```
shouldnt we use XFREE_CLEAR instead ?

teto commented on runtime/doc/eval.txt:
```diff
@@ -5451,6 +5451,9 @@ jobstart({cmd}[, {opts}])				*jobstart()*
 		  |on_exit|  : exit event handler (function name or |Funcref|)
 		  cwd      : Working directory of the job; defaults to
 		             |current-directory|.
+		  env      : A dict of strings to append (or replace see |clear_env|) to the
```
this line is longer than the standard length, add a newline after the closing parenthesis.

teto reviewed (commented):
I've tested on linux successfully
`call jobstart(['sh', '-c', 'echo "test:$TATA with $PATH"' ], {'env': {'TATA': 'SUPER'}, 'clear_env': 1, 'on_stdout':{j,d,e->append(line('.'),d)}})`
I think it's in shape for merging. Ideally we should have a test for the windows scenario you mentioned
```
      // Some Windows env vars start with =, so skip over that to find the
      // separator between name/value
```
but that can come later from a windows guy.

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -26,6 +26,9 @@ describe('jobs', function()
   before_each(function()
     clear()
+
+    assert:set_parameter('TableFormatLevel', 1000000)
```
could be removed since dcc8fcf0b9d819357d5

justinmk commented on src/nvim/eval.c:
```diff
@@ -8717,7 +8717,6 @@ static void f_environ(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
   for (int i = 0; ; i++) {
-    // TODO(justinmk): use os_copyfullenv from #7202 ?
     char *envname = os_getenvname_at_index((size_t)i);
```
@jamessan would it not be better to use `os_copyfullenv` here  and then convert to `rettv->vval.v_dict`? 

Time-cost of calling `os_getenvname_at_index` in a loop is O(n^2).

jamessan replied:
Right, I had thought about that and don't remember why I didn't.

justinmk commented on src/nvim/eval.c:
```diff
@@ -12620,6 +12630,44 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
       }
     }
+    dictitem_T *item;
+    item = tv_dict_find(job_opts, S_LEN("env"));
```
`item` is a rather generic name. could also do assignment at declaration.

justinmk reviewed (approved):
This should have more vigorous tests, but it's sat too long, so let's merge it. Thanks @jamessan @teto

justinmk commented on src/nvim/eval.c:
```diff
@@ -12620,6 +12630,44 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      size_t i = 0;
+      size_t env_size = 0;
+      if (item->di_tv.v_type != VAR_DICT) {
+        EMSG2(_(e_invarg2), "env");
```
missing test

jamessan replied:
You mean you don't like crashes when the type is wrong? :)

jamessan committed:
Add negative test for type of job's env option

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -48,6 +49,57 @@ describe('jobs', function()
+        nvim('command', "let j = jobstart('env', g:job_opts)")
+      end
+    end)
+    ok(string.find(err, "E475: Invalid argument: env") ~= nil)
```
Could use `helpers.pcall_err` next time.

blueyed commented:
This allows for "8.0.0902: cannot specify directory or environment for a job" to be marked as merged, doesn't it?  (https://github.com/vim/vim/commit/05aafed54b50b602315ae55d83a7d089804cecb0)

justinmk commented:
> This allows for "8.0.0902: cannot specify directory or environment for a job" to be marked as merged, doesn't it? ([vim/vim@05aafed](https://github.com/vim/vim/commit/05aafed54b50b602315ae55d83a7d089804cecb0))

@blueyed yes
